### PR TITLE
Legend can have default/constant opacity.

### PR DESCRIPTION
**Pull request by yhoonkim**
FIx https://github.com/vega/vega-lite/issues/2948.

### CODE REVIEW

File: src/compile/legend/encode.ts [hunk 1]
```diff
@@ -60,6 +60,9 @@ export function symbols(fieldDef: FieldDef<string>, symbolsSpec: any, model: Uni
       symbols.shape = {value: shapeDef.value};
     }
   }
+  if (channel !== OPACITY) {
+    symbols.opacity = model.encoding.opacity;
```
Comment (kanitw) [inline src/compile/legend/encode.ts]: It's weird that for shape we have to check if `model.encoding.shape` is a value def but here you don't check the same thing for opacity.

I don't know if the old code is wrong or your new code is wrong, but we definitely need to investigate what's right.  :)

File: src/compile/legend/parse.ts [hunk 2]
```diff
@@ -176,5 +176,21 @@ export function mergeLegendComponent(mergedLegend: LegendComponent, childLegend:
     );
     mergedLegend.setWithExplicit(prop, mergedValueWithExplicit);
   }
+
+  if (mergedLegend.implicit.encode.symbols.update.opacity && mergedLegend.implicit.opacity) {
```
Comment (kanitw) [inline src/compile/legend/parse.ts]: Don't add this HACK here.  

Add the rule inside the `legend/encode.ts` for the symbols method. (If there is no `symbols` method, create one so we have a consistent pattern to determine default values of things in encode block.)

File: src/compile/legend/encode.ts [hunk 3]
```diff
@@ -69,7 +70,18 @@ export function symbols(fieldDef: FieldDef<string>, symbolsSpec: any, model: Uni
   return keys(symbols).length > 0 ? symbols : undefined;
 }
 
-export function labels(fieldDef: FieldDef<string>, labelsSpec: any, model: UnitModel, channel: NonspatialScaleChannel) {
+export function gradient(fieldDef: FieldDef<string>, gradientSpec: any, model: UnitModel, channel: Channel, legendCmpt: LegendComponent) {
+  const legend = model.legend(channel);
+  let gradient:any = {};
+
+  if (legendCmpt.explicit.type === 'gradient' || legendCmpt.implicit.type === 'gradient') {
```
Comment (kanitw) [inline src/compile/legend/encode.ts]: `if (legendCmpt.get('type') === 'gradient')`

File: src/compile/legend/encode.ts [hunk 4]
```diff
@@ -6,9 +6,10 @@ import {keys, without} from '../../util';
 import {applyMarkConfig, timeFormatExpression} from '../common';
 import * as mixins from '../mark/mixins';
 import {UnitModel} from '../unit';
+import {LegendComponent} from './component';
 
 
-export function symbols(fieldDef: FieldDef<string>, symbolsSpec: any, model: UnitModel, channel: Channel) {
+export function symbols(fieldDef: FieldDef<string>, symbolsSpec: any, model: UnitModel, channel: Channel, legendCmpt: LegendComponent) {
```
Comment (kanitw) [inline src/compile/legend/encode.ts]: Why do you need to add a parameter here given you're not modifying the method?

File: test/compile/legend/encode.test.ts [hunk 5]
```diff
@@ -32,6 +32,27 @@ describe('compile/legend', function() {
...
+    });
+  });
+
+  describe('encode.gradient', function() {
+    it('should have default opacity', function() {
+      const gradient = encode.gradient({field: 'a', type: 'quantitative'}, {}, parseUnitModelWithScale({
+          mark: "point",
+          encoding: {
+            x: {field: "a", type: "quantitative"}}
+        }), COLOR, {"implicit":{"type":"gradient"}} as LegendComponent);
```
Comment (kanitw) [inline test/compile/legend/encode.test.ts]: Please don't do unnecessary  `as` here.   You should instead properly create an instance of LegendComponent.

File: examples/compiled/diverging_color_points.vg.json [hunk 6]
```diff
@@ -174,7 +174,23 @@
         {
             "stroke": "color",
             "title": "Weight_in_lbs",
-            "type": "gradient"
+            "type": "gradient",
+            "encode": {
+                "symbols": {
```
Comment (domoritz) [inline examples/compiled/diverging_color_points.vg.json]: we don't need this for gradients

Comment (kanitw) [inline examples/compiled/diverging_color_points.vg.json]: The logic is wrong.  Why would you need to set opacity for both symbols and gradient?

File: examples/compiled/facet_independent_scale_layer.vg.json [hunk 7]
```diff
@@ -506,7 +506,12 @@
     "legends": [
         {
             "stroke": "color",
-            "title": "gender"
+            "title": "gender",
+            "encode": {
```
Comment (kanitw) [inline examples/compiled/facet_independent_scale_layer.vg.json]: Why adding an empty encode block?

File: examples/compiled/interval_mark_style.vg.json [hunk 8]
```diff
@@ -874,7 +874,15 @@
         {
             "fill": "color",
             "title": "Number of Records",
-            "type": "gradient"
+            "type": "gradient",
+            "encode": {
```
Comment (kanitw) [inline examples/compiled/interval_mark_style.vg.json]: Empty block again.

File: src/compile/legend/encode.ts [hunk 9]
```diff
@@ -1,14 +1,19 @@
...
 import {keys, without} from '../../util';
 import {applyMarkConfig, timeFormatExpression} from '../common';
 import * as mixins from '../mark/mixins';
 import {UnitModel} from '../unit';
+import {LegendComponent} from './component';
 
 
-export function symbols(fieldDef: FieldDef<string>, symbolsSpec: any, model: UnitModel, channel: Channel) {
+export function symbols(fieldDef: FieldDef<string>, symbolsSpec: any, model: UnitModel, channel: Channel, legendCmpt: LegendComponent) {
+  if (((legendCmpt.explicit || {}).type === 'gradient') || ((legendCmpt.implicit || {}).type === 'gradient')){
```
Comment (kanitw) [inline src/compile/legend/encode.ts]: `legendCmpt.get('type') ===  'gradient'`

(Maybe you miss my comment earlier)

File: src/compile/legend/encode.ts [hunk 10]
```diff
@@ -61,12 +66,33 @@ export function symbols(fieldDef: FieldDef<string>, symbolsSpec: any, model: Uni
...
   symbols = {...symbols, ...symbolsSpec};
 
   return keys(symbols).length > 0 ? symbols : undefined;
 }
 
-export function labels(fieldDef: FieldDef<string>, labelsSpec: any, model: UnitModel, channel: NonspatialScaleChannel) {
+export function gradient(fieldDef: FieldDef<string>, gradientSpec: any, model: UnitModel, channel: Channel, legendCmpt: LegendComponent) {
+  let gradient:any = {};
+
+  if (((legendCmpt.explicit || {}).type === 'gradient') || ((legendCmpt.implicit || {}).type === 'gradient')) {
```
Comment (kanitw) [inline src/compile/legend/encode.ts]: legendCmpt.get('type') === 'gradient'

File: test/compile/legend/encode.test.ts [hunk 11]
```diff
@@ -58,7 +88,7 @@ describe('compile/legend', function() {
       });
 
       const fieldDef = {field: 'a', type: TEMPORAL, timeUnit: TimeUnit.QUARTER};
-      const label = encode.labels(fieldDef, {}, model, COLOR);
+      const label = encode.labels(fieldDef, {}, model, COLOR,  legendCompt);
```
Comment (kanitw) [inline test/compile/legend/encode.test.ts]: one space after comma is sufficient

File: test/compile/legend/parse.test.ts [hunk 12]
```diff
@@ -37,6 +40,12 @@ describe('compile/legend', function() {
         const model = parseUnitModelWithScale(s);
 
         const def = legendParse.parseLegendForChannel(model, channel).combine();
+
+        if (channel !== OPACITY){
+          assert.equal(def.encode.symbols.update.opacity.value, "0.7");
```
Comment (kanitw) [inline test/compile/legend/parse.test.ts]: It should be `0.7`, not `"0.7"`?

File: src/util.ts [hunk 13]
```diff
@@ -238,3 +238,18 @@ export function logicalExpr<T>(op: LogicalOperand<T>, cb: Function): string {
 // Omit from http://ideasintosoftware.com/typescript-advanced-tricks/
 export type Diff<T extends string, U extends string> = ({[P in T]: P } & {[P in U]: never } & { [x: string]: never })[T];
 export type Omit<T, K extends keyof T> = {[P in Diff<keyof T, K>]: T[P]};
+
+// delete an object's empty key until it is not empty.
```
Comment (kanitw) [inline src/util.ts]: /**
 * Delete nested property of an object, and delete the ancestors of the property if they become empty.
 */ 

Name it `deleteNestedProperty`.

**Pull request merged**